tests/source/chains_long_items_block.rs

```diff
@@ -0,0 +1,175 @@
+}
+
+// https://github.com/rust-lang/rustup.rs/pull/2097
+fn deeply_nested() {
+    let mut app = App::new("rustup")
+    .version(common::version())
+    .about("The Rust toolchain installer")
+    .after_help(RUSTUP_HELP)
+    .setting(AppSettings::VersionlessSubcommands)
+    .setting(AppSettings::DeriveDisplayOrder)
+    .setting(AppSettings::SubcommandRequiredElseHelp)
+.subcommand(
+    SubCommand::with_name("run")
+        .about("Run a command with an environment configured for a given toolchain")
+        .after_help(RUN_HELP)
+        .setting(AppSettings::TrailingVarArg)
+        .arg(
+            Arg::with_name("install")
+.help("Install the requested toolchain if needed")
+.long("install"),
+        )
+            .arg(
+                Arg::with_name("toolchain")
+                            .help(TOOLCHAIN_ARG_HELP)
+        .required(true),
+            )
+            .arg(
+        Arg::with_name("command")
+            .required(true)
+            .multiple(true)
+            .use_delimiter(false),
+            ),
+    )
+    .subcommand(
+        SubCommand::with_name("which")
+    .about("Display which binary will be run for a given command")
+    .arg(Arg::with_name("command").required(true))
+    .arg(
+        Arg::with_name("toolchain")
+            .help(TOOLCHAIN_ARG_HELP)
+            .long("toolchain")
+            .takes_value(true),
+    ),
+    )
+    .subcommand(
+        SubCommand::with_name("doc")
+            .alias("docs")
+            .about("Open the documentation for the current toolchain")
+            .after_help(DOC_HELP)
+            .arg(
+Arg::with_name("path")
+                            .long("path")
+                    .help("Only print the path to the documentation"),
+    )
+    .args(
+        &DOCS_DATA
+            .iter()
+            .map(|(name, help_msg, _)| Arg::with_name(name).long(name).help(help_msg))
+            .collect::<Vec<_>>(),
+    )
+    .arg(Arg::with_name("toolchain")
+            .help(TOOLCHAIN_ARG_HELP)
+            .long("toolchain")
+            .takes_value(true),
+    ).group(ArgGroup::with_name("page").args(
+            &DOCS_DATA
+                .iter()
+                .map(|(name, _, _)| *name)
+                .collect::<Vec<_>>(),
+        ),
+    )
+    .arg(Arg::with_name("topic").help("Topic such as 'core', 'fn', 'usize', 'eprintln!', 'core::arch', 'alloc::format!', 'std::fs', 'std::fs::read_dir', 'std::io::Bytes', 'std::iter::Sum', 'std::io::error::Result' etc..."),
+        ),
+    );
+}
+
+fn inner_chain() {
+    let foo = bar().asdffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff(
+Foo {
+    field1: val1,
+    field2: val2,
+}.method_call().method_call(),
+    ).foo().bar().baz();
+}
+
+fn long_parent() {
```

Comment:
While working on this I discovered quite a variety of chain scenarios (position of long element in the chain, rhs of an assignment vs. free, etc.) that had to be explicitly handled, so there's test cases for each of those scenarios